### ISSUE TITLE
ca-certificates.crt is deleted on postfix restart

### ISSUE THREAD

**Issue by Wasca** [tarickb/sasl-xoauth2#13]
I've found that Postfix is deleting /var/spool/postfix/etc/ssl/certs/ca-certificates.crt when you restart the Postfix service using the command...

`service postfix restart`

When the file is missing from that location you the token does not get refreshed and you can no longer send email you get this error.

`status=deferred (SASL authentication failed; cannot authenticate to server smtp.gmail.com[108.177.97.108]: bad protocol / cancel)`

It appears to be this script getting called (/usr/lib/postfix/configure-instance.sh) during the restart script that causes all the trouble. 

Is this something that is meant to happen?

**Comment by Wasca** [comment]
I'm not sure if this is correct way to fix this but I just added this line before the closing `fi` in `/usr/lib/postfix/configure-instance.sh`

` cp /etc/ssl/certs/ca-certificates.crt /var/spool/postfix/etc/ssl/certs/ca-certificates.crt`

Each time postfix is started or restarted the certificate is now copied into the correct location.

I need to find a way to execute this copy command in the postfix startup script, I tried but it never worked when I entered it in there.

**Comment by dberlin** [comment]
The more "correct" patch is that in configure-instance.sh, change where you see -name '*.pem' to -name '*.pem' -or -name '*.crt'

**Comment by rrthomas** [comment]
FWIW, on my Ubuntu 20.04 machine, I have found that the certificates file is *not* deleted when I restart postfix. From some of the above comments, I infer that this only happens with `postfix-multi` installations; can anyone confirm or deny this?

In PR #25 I add support for automatically updating the certificates for the default single-instance case. A multi-instance user might be interested in extending this, taking into account the deletion of the certificates on restart, if that still occurs.

**Comment by daniel-fahey** [comment]
I had the same issue with Postfix recreating the chroot and wiping `ca-certificates.crt` every time it (re)starts on Debian/Ubuntu. Found a workaround by tweaking the Postfix service to run `update-ca-certificates` before it (re)starts:

1. Run `sudo systemctl edit postfix.service`
2. Add these lines to the file:
```
[Service]
ExecStartPre=/usr/sbin/update-ca-certificates
```
3. Save it, then run `sudo systemctl daemon-reload` and `sudo systemctl restart postfix`.

Of course you'd need the `/etc/ca-certificates/update.d/postfix-sasl-xoauth2-update-ca-certs` installed.

This fixed it for me, hope it helps! Anyone know how this might be included in the Debian/Ubuntu package?

**Comment by rrthomas** [comment]
To get it into Debian/Ubuntu, file a Debian bug with as much help for the package maintainer to fix it as possible.

**Comment by mjt0k** [comment]
FWIW, I think the only solution to this in postfix (in debian) is to un-chroot the service in question.  There's no way we can keep chroot updated with everything which is needed for various possible SASL or nsswitch or other plugins.

**Comment by mjt0k** [comment]
speaking of /etc/ssl/certs/ca-certificates.crt -- can't the software pick the hashed certs which are readily available in the same dir?  The ca-certificates.crt is basically a combination of all the *.pem files in there, which are also in [0-9a-f]{8}\.[0-9] files.  When verifying a certificate, it shouldn't need whole thing, it is enough to look at $hash.N files.

**Comment by mjt0k** [comment]
https://docs.openssl.org/master/man3/SSL_CTX_load_verify_locations/ describes several functions which can be used to point libssl to the desired certs locations.  It feels like xoauth2 should use `SSL_CTX_set_default_verify_store()` routine instead of `SSL_CTX_load_verify_file()`.

**Comment by minusdavid** [comment]
Having this same problem with sasl-xoauth2 0.27-1ubuntu1~noble1~ppa1 on Ubuntu 24.04.3 where /var/spool/postfix/etc/ssl/certs/ca-certificates.crt gets deleted on every Postfix restart.

**Comment by tarickb** [comment]
@minusdavid do you have `smtp_tls_CAfile = /etc/ssl/certs/ca-certificates.crt` set in your `/etc/postfix/main.cf`?

**Comment by mjt0k** [comment]
just un-chroot your postfix daemons, instead of jumping through the hoops trying to work with it chrooted.  In next release of debian (and hopefully ubuntu will follow), postfix will not be run chrooted, - finally, after 25 years of constant grief and struggling.

**Comment by minusdavid** [comment]
> [@minusdavid](https://github.com/minusdavid) do you have `smtp_tls_CAfile = /etc/ssl/certs/ca-certificates.crt` set in your `/etc/postfix/main.cf`?

No, that line is commented out.



> just un-chroot your postfix daemons, instead of jumping through the hoops trying to work with it chrooted. In next release of debian (and hopefully ubuntu will follow), postfix will not be run chrooted, - finally, after 25 years of constant grief and struggling.

I tried to un-chroot it back in January, but I must not have restarted the right processes, as I couldn't get the config to take. It's a good idea though. I'll try that next time I'm setting up.

**Comment by tarickb** [comment]
> > [@minusdavid](https://github.com/minusdavid) do you have `smtp_tls_CAfile = /etc/ssl/certs/ca-certificates.crt` set in your `/etc/postfix/main.cf`?
> 
> No, that line is commented out.

Try enabling it? Last I checked, if `smtp_tls_CAfile` is set, Postfix ensures the CA file is in the chroot jail when it starts.

**Comment by minusdavid** [comment]
I'll give it a go next time I'm trying to set it up. I did a test a while back with Google credentials and currently waiting on some Microsoft credentials to do a real implementation.

**Comment by mirolm** [comment]
You can pretty much skip the script that triggers with update-ca-certificates as it only causes confusion and token renewal breaks on next machine/postfix restart.

Below worked better for me.

sudo systemctl edit postfix.service

````
[Service]
ExecStartPost=+/usr/bin/mkdir -p /var/spool/postfix/etc/ssl/certs
ExecStartPost=+/usr/bin/cp /etc/ssl/certs/ca-certificates.crt /var/spool/postfix/etc/ssl/certs/ca-certificates.crt
````

**Comment by mjt0k** [comment]
Once again, please do  not add extra work-arounds for a non-issue.  Just un-chroot your postfix and be done with it.  It was wrong decision to chroot it by default in early 2000s, everyone in the postfix community pointed this out countless number of times, and it caused a lot of damage over 2 decades.  There's no excuse to add more bandaids, just do the right thing and be done with it.  In current debian, you can run `postfix chroot off` (and `postfix reload`) to do that.

It is very sad people offer more and more wrong bandaids instead of following the right solution.

**Comment by minusdavid** [comment]
Yeah I just unchroot it these days
